### DISCUSSION TITLE
How not to provide the keys for "deep" types every time?

### DISCUSSION THREAD

**Discussion by ProdigySim** [millsp/ts-toolbelt#190] [Wiki]
#### Is your feature request related to a problem?
No.

#### Describe the solution you'd like
```ts
type DeepOptional<O extends object, K extends Index = keyof O> = Optional<O, K, 'deep'>'
```
Using `Object.Optional` is very handy. But if you want to use the `'deep'` option, it requires filling in the optional `K` parameter as well. For most cases where I want to use a deep-optional object, I'd like to use the default `K = keyof T` argument, so it's inconvenient to have to specify this parameter manually. Creating an alias for the `'deep'` version of the command would allow people to leverage the nice defaults available.

#### Describe alternatives you've considered
I believe Typescript has discussed a feature where you could use `_` in place of a generic parameter to accept the default value. This would let users write `O.Optional<T, _, 'deep'>` instead. However, I can no longer find the TS issue discussing this feature, so I don't know when or if it is expected to arrive.

Additionally, users could just create their own `DeepOptional` alias. But it has been nice to be able to consolidate many type aliases in a standard way under a single library like ts-toolbelt.

#### Teachability, Documentation, Adoption, Migration Strategy
We should be able to copy the documentation for `Object.Optional`, adding a note that this method is an alias for `Object.Optional<O,K, 'deep'>` and functionality is the same.

#### Examples
```ts
declare const mockItem: O.Optional<IItem, keyof IItem, 'deep'>;
// vs
declare const mockItem: O.DeepOptional<IItem>;
```

**Comment by millsp** [comment]
Hi @ProdigySim,

This is something that is not well documented, but you can do:

```ts
type test = O.Optional<IItem, any, 'deep'>;
```

Is that what you were looking for?

**Comment by ProdigySim** [comment]
Ah, that's definitely a lot better than having to duplicate the first type parameter. I think an alias could still be nice, but the `any` workaround does remove a lot of the need.

**Comment by millsp** [comment]
I'm marking this as resolved, as on of my goals is to provide a unified API for the type utilities (no duplicates). I feel like typing `any` is short enough not to create a new alias.

Cheers

**Comment by glen-84** [comment]
> However, I can no longer find the TS issue discussing this feature

https://github.com/microsoft/TypeScript/issues/26242
